src/ai/backend/manager/scheduler/dispatcher.py

```diff
@@ -623,6 +626,27 @@ async def _update_session_status_data() -> None:
         if num_scheduled > 0:
             await self.event_producer.produce_event(DoPrepareEvent())
 
+    async def _filter_agent_by_container_limit(
+        self, candidate_agents: list[AgentRow]
+    ) -> list[AgentRow]:
+        raw_value = await self.shared_config.etcd.get("config/agent/max-container-count")
+        if raw_value is None:
+            return candidate_agents
+        max_container_count = int(raw_value)
+
+        async def _pipe_builder(r: Redis) -> RedisPipeline:
+            pipe = r.pipeline()
+            for ag in candidate_agents:
+                await pipe.get(f"container_count.{ag.id}")
+            return pipe
+
+        raw_counts = await redis_helper.execute(self.registry.redis_stat, _pipe_builder)
+        return [
+            ag
+            for ag, count in zip(candidate_agents, raw_counts)
+            if max_container_count > int(count)
```

Comment:
It would be safer to check if `count is not None` for cases when there are missing results when querying `container_count.{agent_id}` keys in Redis.

Reply:
thanks, I forgot that the result from redis could be `None`.
And I add `container_count` field to gql object.